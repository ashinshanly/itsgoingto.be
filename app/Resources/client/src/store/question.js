import { createSelector } from 'reselect'
import { compose, not, equals, length } from 'ramda'
import { addAnswer, removeAnswers } from './answers'

// ------------------------------------
// Constants
// ------------------------------------
export const QUESTION_UPDATE = 'QUESTION_UPDATE'

// ------------------------------------
// Selectors
// ------------------------------------
export const questionSelector = (state) => state.question

export const hasQuestionSelector = createSelector(
  questionSelector,
  question => compose(not, equals(0), length)(question)
)

// ------------------------------------
// Actions
// ------------------------------------
export const updateQuestion = (value = '') => (dispatch, getState) => {
  let hadQuestion = hasQuestionSelector(getState())
  dispatch ({
    type    : QUESTION_UPDATE,
    text : value
  })
  let hasQuestion = hasQuestionSelector(getState())

  if (hasQuestion && !hadQuestion) {
    dispatch (addAnswer())
  } else if (hadQuestion && !hasQuestion) {
    dispatch (removeAnswers())
  }
}

export const actions = {
  updateQuestion
}

// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  [QUESTION_UPDATE] : (previousState, action) => action.text
}

// ------------------------------------
// Reducer
// ------------------------------------
const initialState = ''
export default function questionReducer (state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type]

  return handler ? handler(state, action) : state
}
