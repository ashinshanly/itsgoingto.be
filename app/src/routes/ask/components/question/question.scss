.question-container {
  transition: top 0.4s ease;
  top: 0px;
  &.move-up {
    top: -10em;
  }

  .input-question {
    margin-bottom: 1.3em;

    .input-field-question {
      font-size: 2em;
    }
  }
}
