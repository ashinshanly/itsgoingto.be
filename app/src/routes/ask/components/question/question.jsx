import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { compose, nth, slice, concat, __, when, merge, ifElse, add, equals, subtract, length } from 'ramda'
import autosize from 'autosize'
import { updateQuestion } from 'store/poll'
import EventBus from 'components/event-bus'
import './question.scss'

const KEY_DOWN_ARROW = 40
const KEY_ENTER = 13

class Question extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      placeholder : 0,
      character : 0,
      cursor : false,
    }
  }

  handleChange = (event) => {
    const { onQuestionChange } = this.props

    onQuestionChange(event.target.value)
  }

  handleKeyPress = (event) => {
    event = event || window.event
    const key = event.keyCode || event.charCode

    switch (key) {
      case KEY_DOWN_ARROW:
      case KEY_ENTER:
        event.preventDefault()
        this.eventBus.emit('focus', 0)
        break
    }
  }

  humanize = () => Math.round(Math.random() * (150 - 30)) + 30

  toggleCursor = () => {
    this.setState((prevState) =>
      merge(prevState, { cursor : !prevState.cursor })
    )
  }

  type = () => {
    this.setState((prevState) =>
      merge(prevState, { character : prevState.character + 1 })
    )
    this.characterUpdate = setTimeout(
      this.type,
      this.humanize()
    )
    if (this.state.character <= this.props.placeholderText[this.state.placeholder].length) {
      clearInterval(this.cursorUpdater)
      this.setState(merge(this.state, { cursor : true }))
      this.cursorUpdater = setInterval(
        this.toggleCursor,
        500
      )
    }
  }

  updatePlaceholder = () => {
    this.setState((prevState, props) =>
      merge(prevState, {
        character   : 0,
        placeholder : ifElse(
          equals(compose(subtract(__, 1), length)(props.placeholderText)),
          () => 0,
          add(1)
        )(prevState.placeholder)
      })
    )
  }

  placeholderSelector = (state) =>
    compose(
      when(
        () => state.cursor,
        concat(__, '|')
      ),
      slice(0, state.character),
      nth(state.placeholder)
    )(this.props.placeholderText)

  componentDidMount = () => {
    this.eventBus = EventBus.getEventBus()
    this.eventListener = this.eventBus.addListener('focus', (index) => {
      if (index === -1) {
        this.refs.question.focus()
      }
    })
    autosize(this.refs.question)
    this.cursorUpdater = setInterval(
      this.toggleCursor,
      500
    )
    this.characterUpdate = setTimeout(
      this.type,
      this.humanize()
    )
    this.placeholderUpdater = setInterval(
      this.updatePlaceholder,
      5000
    )
  }

  componentWillUnmount = () => {
    this.eventListener.remove()
    autosize.destroy(this.refs.question)
    clearInterval(this.cursorUpdater)
    clearTimeout(this.characterUpdate)
    clearInterval(this.placeholderUpdater)
  }

  render () {
    const { question } = this.props

    return (
      <div className='input input-question'>
        <label className='input-label input-label-question' htmlFor='question'>Ask a question</label>
        <textarea
          className='input-field input-field-question js-auto-size'
          value={question}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyPress}
          placeholder={this.placeholderSelector(this.state)}
          rows='1'
          id='question'
          name='question'
          ref='question' />
      </div>
    )
  }
}

Question.propTypes = {
  placeholderText  : PropTypes.array.isRequired,
  question         : PropTypes.string.isRequired,
  onQuestionChange : PropTypes.func.isRequired
}

const mapDispatchToProps = (dispatch) => ({
  onQuestionChange : (value) => dispatch(updateQuestion(value))
})

export default connect(null, mapDispatchToProps)(Question)
