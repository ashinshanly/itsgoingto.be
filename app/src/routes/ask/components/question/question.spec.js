/* global expect, jest */
import React from 'react'
import { shallow } from 'enzyme'
import EventBus from 'components/event-bus'
import { Question } from './question'

const eventBus = {
  emit        : jest.fn(),
  addListener : jest.fn(),
}

EventBus.getEventBus = jest.fn(() => eventBus)

const props = {
  placeholderText: [
    'Ask a question?',
    'Ask a different question?',
  ],
  question         : '',
  onQuestionChange : jest.fn(),
}

describe('(Route) Ask', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('(Component) Question', () => {
    describe('(Action) onChange', () => {
      it('should call prop onQuestionChange', () => {
        const wrapper = shallow(<Question {...props} />)
        const event = {
          preventDefault() { },
          target: { value: 'A question' }
        }

        wrapper.find('textarea').simulate('change', event)
        expect(props.onQuestionChange).toHaveBeenCalledWith('A question')
      })
    })

    describe('(Action) onKeyDown', () => {
      const emitFocusPlusZero = (keyCode) => {
        it('emits focus request on index + 1', () => {
          const wrapper = shallow(<Question {...props} />)

          wrapper.find('textarea').simulate('keyDown', { keyCode, preventDefault: () => { } })
          expect(eventBus.emit).toHaveBeenCalledWith('focus', 0)
        })
      }

      describe('key is KEY_DOWN_ARROW', () => {
        emitFocusPlusZero(40)
      })

      describe('key is KEY_ENTER', () => {
        emitFocusPlusZero(13)
      })
    })

    describe('(Render)', () => {
      it('matches snapshot', () => {
        const wrapper = shallow(<Question {...props} />)
        expect(wrapper).toMatchSnapshot()
      })

      describe('with question text', () => {
        it('matches snapshot', () => {
          const wrapper = shallow(<Question {...props} question={'A real question'} />)
          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('with place holder text', () => {
        it('matches snapshot', () => {
          const wrapper = shallow(<Question {...props} />)
          wrapper.setState({
            placeholder : 1,
            character   : 12,
            cursor      : true,
          })
          expect(wrapper).toMatchSnapshot()
        })
      })
    })
  })
})
