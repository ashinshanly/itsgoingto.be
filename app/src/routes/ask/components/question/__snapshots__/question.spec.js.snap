// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Route) Ask (Component) Question (Render) matches snapshot 1`] = `
<div
  className="input input-question"
>
  <label
    className="input-label input-label-question"
    htmlFor="question"
  >
    Ask a question
  </label>
  <textarea
    className="input-field input-field-question js-auto-size"
    id="question"
    name="question"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder=""
    rows="1"
    value=""
  />
</div>
`;

exports[`(Route) Ask (Component) Question (Render) with place holder text matches snapshot 1`] = `
<div
  className="input input-question"
>
  <label
    className="input-label input-label-question"
    htmlFor="question"
  >
    Ask a question
  </label>
  <textarea
    className="input-field input-field-question js-auto-size"
    id="question"
    name="question"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Ask a differ|"
    rows="1"
    value=""
  />
</div>
`;

exports[`(Route) Ask (Component) Question (Render) with question text matches snapshot 1`] = `
<div
  className="input input-question"
>
  <label
    className="input-label input-label-question"
    htmlFor="question"
  >
    Ask a question
  </label>
  <textarea
    className="input-field input-field-question js-auto-size"
    id="question"
    name="question"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder=""
    rows="1"
    value="A real question"
  />
</div>
`;
